Implement `Tracker`

p0lunin opened the pull request:
This Pull Request adds a `Tracker` part of the mint. The `Tracker` responsible for keeping track of new bitcoin transactions and detecting mint bids among them. 

- Process verified bitcoin blocks, detect OP_RETURN bids.
  - [x] Developed.
  - [x] Tested with bitcoin node.
- Detect bitcoin reorgs.
  - [x] Developed.
  - [x] Tested with bitcoin node.
    - [x] Fork where longer chain is main.
    - [x] ~~Fork where shorter chain is main.~~ This is untrivial, see 6c89c91.
- Submit bid transactions.
  - [x] Developed.
  - [x] Tested with bitcoin node.
- Index bids for quick reference when validating sidechain blocks
  - [x] Developed.
  - [x] Tested with bitcoin node.

p0lunin committed:
adds Tracker struct with check_bitcoin_block_with_id method

p0lunin committed:
refactored check_bitcoin_block_with_id: splited at 3 functions

p0lunin committed:
added inserting transactions into indexes and checks for reorgs

p0lunin committed:
added inserting transactions into indexes and checking for reorgs

p0lunin committed:
added BitcoinClient struct, TestClient for testing and tests for Index struct

p0lunin committed:
fmt

p0lunin committed:
move tracker into separate crate

p0lunin committed:
small fixes

p0lunin committed:
wip

p0lunin committed:
added bag_ids to tracker

p0lunin committed:
fixed some errors in the tests

p0lunin committed:
refactoring + fmt

p0lunin committed:
refactored Index to be more readable

p0lunin committed:
added function to create mint transaction in the bitcoin chain

p0lunin committed:
prettified bitcoin node test

p0lunin committed:
refactoring

p0lunin committed:
added `Storage` trait for store index records

p0lunin committed:
added part of sqlite storage functions [WIP]

p0lunin committed:
added implementation of remove_with_block_hash method

p0lunin committed:
removed writing by BLOB api

p0lunin committed:
added `bundled` feature to rusqlite

p0lunin committed:
made sqlite module public

p0lunin committed:
added test with bitcoin node for sqlite storage

p0lunin committed:
types from `storage` separated at other files

oleganza commented on tracker/src/index.rs:
```diff
@@ -0,0 +1,331 @@
+        match self.height.cmp(&new_height) {
+            Ordering::Equal => return,
+            Ordering::Greater => {
+                unimplemented!("TODO: is it possible that chain will be less than previous?")
```
It could be in case of a deep or across-retarget reorg. Difficulty and each block’s weight is fixed within each retarget interval (2016 blocks, ≈2 wks). So within an interval, with common ancestor in the same interval, a chain with lower height is strictly lower in weight. Otherwise, not.

p0lunin replied:
Probably resolved in a9fd259, but tests are needed.

p0lunin replied:
It is hard to test it using bitcoin node 6c89c91, so I added test using mock version of bitcoin node: a9a89fd

oleganza commented on tracker/src/index.rs:
```diff
@@ -0,0 +1,331 @@
+        let push_bytes_instr = instructions.next().and_then(|res| res.ok())?;
+        let bag_id = parse_push_32_bytes(push_bytes_instr)?;
+
+        if !self.bags.iter().any(|bag| *bag == bag_id) {
```
Is it possible to have bitcoin block scanned before the bags become known? Imagine we are bootstrapping the node and ingesting both chains. One network might be faster than the other.

p0lunin replied:
Resolved by https://github.com/runflame/mint/pull/1/commits/3f74bd4a8d30cbd8f788f02831f9599560ad8a89.

p0lunin committed:
added `btc` suffix to names that related to bitcoin chain

p0lunin committed:
removed unused

p0lunin committed:
added remove_records_with_bag for IndexStorage

p0lunin committed:
added create_test_mint_transaction fn

p0lunin committed:
now bags in btc transactions must be checked manually by validate_bags fn

p0lunin committed:
rename get_blocks_by_hash -> get_records_by_block_hash

p0lunin committed:
added shutdown bitcoind processes after tests

p0lunin committed:
fmt

p0lunin committed:
removed TODO comment

oleganza commented on tracker/tests/setup_single_node.sh:
```diff
@@ -0,0 +1,24 @@
+PORT=$1
```
Could you please add an example in README on how to run this script?

p0lunin replied:
I have added description and examples in 94c56bb59e24c4cd880602d7581c29d4a8b9eaf8. Now is it clear now how to use the script?

oleganza replied:
Great, thank you.

p0lunin committed:
added docs how to use setup_single_node.sh

p0lunin committed:
move utils into utils.rs

p0lunin committed:
method for getting btc client from tracker

p0lunin committed:
added few methods

p0lunin committed:
add reorg test with longest chain

p0lunin committed:
fmt

p0lunin committed:
refactoring

p0lunin committed:
change format for the mint transactions to p2wsh

p0lunin committed:
up amount of the transactions

p0lunin committed:
Full rework
1. Tracker now can add bids using "proofs" - bitcoin outpoint + bag id (it does not verify bag itself because it's not their job).
2. Internal indexing process was simplified: removed useless caches and variables.
3. Bags are stored in the storage, not in the vector.
4. Tests are reworked too.

p0lunin committed:
rename BagEntry -> BidEntry

p0lunin committed:
Changes for bag storage:
1. Added api for confirmed|unconfirmed bags.
2. Renamed BagHashSetStorage -> BagMemoryStorage.
3. Moved memory store into another module.

p0lunin committed:
change Index functions to use api of confirmed|unconfirmed bags

p0lunin committed:
fmt

p0lunin committed:
rename IndexStorage -> BidStorage

p0lunin committed:
Remove gettransaction method as it requires the txid to belong to the wallet.

p0lunin committed:
Add #[allow(unused)]

p0lunin committed:
update tests

p0lunin committed:
Remove test with reorg on shorter chain in bitcoin tests, because reorg on shorter chain requires to have a higher difficulty on shorter chain, but it cannot be set in generating methods https://developer.bitcoin.org/reference/rpc/#generating-rpcs, so it requires to mine a block on test app side, which is not the task we want to do in test function.

p0lunin committed:
Fix bug with reorg on shorter chain.

p0lunin committed:
Add test with reorg on shorter chain.

p0lunin committed:
merge BagStorage with BidStorage

p0lunin committed:
added error handling

p0lunin committed:
make newtype BagId

p0lunin committed:
impl Display for BagId

p0lunin committed:
added error handling in get_records_by_block_hash instead of panicking

p0lunin committed:
documentation and renaming

p0lunin committed:
rename remove_with_block_hash -> remove_confirmation_with_block_hash, also changed implementation: now it set bag status to "unconfirmed"

p0lunin committed:
remove restrictions on `Index` generics

p0lunin committed:
Merge branch 'main' of https://github.com/runflame/mint into main
